demos/multi_material/compositional_buoyancy/van_keken_1997.py

```diff
@@ -0,0 +1,276 @@
+# measure of distance from that interface. In the conservative level-set approach, the
+# classic signed-distance function, $\phi$, employed in the level-set method is
+# transformed into a smooth step function, $\psi$, according to
+#
+# $$\psi(\mathbf{x}, t) = \frac{1}{2} \left[
+# \mathrm{tanh} \left( \frac{\phi(\mathbf{x}, t)}{2\epsilon} \right) + 1
+# \right]$$
+#
+# Throughout the simulation, the level-set field is advected with the flow:
+#
+# $$\frac{\partial \psi}{\partial t} + \nabla \cdot \left( \mathbf{u}\psi \right) = 0$$
+#
+# Advection of the level set modifies the shape of the initial profile. In other words,
+# the signed-distance property underpinning the smooth step function is lost. To
+# maintain the original profile as the simulation proceeds, a reinitialisation
+# procedure is employed. We choose the equation proposed in Parameswaran and Mandal
+# (2023):
+#
+# $$\frac{\partial \psi}{\partial \tau_{n}} = \theta \left[
+# -\psi \left( 1 - \psi \right) \left( 1 - 2\psi \right)
+# + \epsilon \left( 1 - 2\psi \right) \lvert\grad\psi\rvert
+# \right]$$
+#
+# This example
+# ------------
+#
+# Here, we consider the isoviscous Rayleigh-Taylor instability presented in van Keken
+# et al. (1997). Inside a 2-D domain, a buoyant, lighter material sits beneath a denser
+# material. The initial material interface promotes the development of a rising
+# instability on the domain's left-hand side, and further smaller-scale convective
+# dynamics take place throughout the remainder of the simulation.
+
+# As with all examples, the first step is to import the gadopt module, which
+# provides access to Firedrake and associated functionality.
+
+from gadopt import *
+
+# We next set up the mesh and function spaces and specify functions to hold our
+# solutions, as in our previous tutorials.
+
+# +
+nx, ny = 40, 40  # Number of cells in x and y directions
+lx, ly = 0.9142, 1  # Domain dimensions in x and y directions
+# Rectangle mesh generated via Firedrake
+mesh = RectangleMesh(nx, ny, lx, ly, quadrilateral=True)
+left_id, right_id, bottom_id, top_id = 1, 2, 3, 4  # Boundary IDs
+
+V = VectorFunctionSpace(mesh, "CG", 2)  # Velocity function space (vector)
+W = FunctionSpace(mesh, "CG", 1)  # Pressure function space (scalar)
+Z = MixedFunctionSpace([V, W])  # Stokes function space (mixed)
+Q = FunctionSpace(mesh, "CG", 2)  # Temperature function space (scalar)
+K = FunctionSpace(mesh, "DQ", 2)  # Level-set function space (scalar, discontinuous)
+R = FunctionSpace(mesh, "R", 0)  # Real space for time step
+
+z = Function(Z)  # A field over the mixed function space Z
+u, p = split(z)  # Symbolic UFL expressions for velocity and pressure
+z.subfunctions[0].rename("Velocity")  # Associated Firedrake velocity function
+z.subfunctions[1].rename("Pressure")  # Associated Firedrake pressure function
+T = Function(Q, name="Temperature")  # Firedrake function for temperature
```

Comment:
Does this need to be Q2? Could we not put it on CG 1 (W) function space, thus reducing the need for Q above (given that T is not really necessary for this simulation, aside from G-ADOPT requirement!)

Reply:
That will not change the fact that `T` is there. Defining `Q` aligns with other tutorials.

Reply:
True - but it's not really used here, hence confusing. Happy to leave as is for now and come back to this down the track if we find a way to not require T for simple multi-material Stokes solves.